network-rfc/12_contracts_design.md

```diff
@@ -0,0 +1,90 @@
+# Portal System
+
+
+### Portal Creation and SQD Collection
+
+The Portal System begins when a Data Consumer creates a new portal through the PortalFactory by specifying key parameters including the target amount of SQD tokens needed, deposit deadline, payment token, "budget".
+
+The data consumer allocation (contribution by the deployer) will be determined by the target amount that the data consumer is seeking.
+
+We are collecting 120% of the amount that will be set by SQD.
+
+The factory deploys a single PortalProxy contract, an upgradeable instance that combines both the core distribution logic and SQD vault functionality into one unified contract. 
+Once deployed, SQD token providers can stake their tokens directly into the PortalProxy by calling the stake function with the portal address and desired amount. 
+
+During this collection phase, the portal remains in a "Collecting" state where it accumulates SQD deposits from multiple providers until either the target amount is reached or the deposit deadline passes. 
+If the target is met before the deadline, the data consumer can trigger the activate function to transition the portal to its active distribution phase.
+
+However, if the deadline expires without reaching the target, the portal is marked as failed, triggering a full refund of both the consumer's budget and all staked SQD tokens back to their respective owners.
+
+### Active Distribution and Fee Routing
+Once activated, the portal enters its Active state where it begins distributing 
+
+Throughout this active period, the data consumer can call the distribute function to inject tokens into the contract, which will be distributed across SQD providers etc. 
+This amount is  distributed based on the FeeRouterModule, a separate admin-controlled contract responsible for splitting the fees according to configurable basis point allocations (defaulting to 50% for SQD providers, 45% for the worker reward pool, and 5% for burning).
+
+The FeeRouterModule holds the actual BPS.
+ During both the staking and distribution phases, the system can trigger external Hooks at key moments (before and after staking, distribution, and exits), allowing for customized behavior such as  additional protocol token rewards layered on top of base distributions etc.
+ Similar to UniswapV4 Hooks.
+
+### Reward Claims, Exits, and Closure
+
+While the portal is active, SQD providers can claim their proportional share of accumulated rewards at any time by calling the claimRewards function on the PortalProxy, which calculates their share based on their staked balance relative to the total tokens in the portal and transfers the corresponding tokens to them. The portal continues distributing as long as the data consumer injects tokens through the distribute function, with all distributions based on the FeeRouterModule configured splits.
```

Comment:
Does the `claimRewards` function calculate their share? I think you won't be able to account for past clamis this way, so you have to top up claimable amount at the distribution time. The same is already done in our WorkerRegistration contract